sdk/azidentity/azure_developer_cli_credential.go

```diff
@@ -0,0 +1,170 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azidentity
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+const (
+	credNameAzureDeveloperCLI = "AzureDeveloperCLICredential"
+	timeoutAzdRequest         = 10 * time.Second
+)
+
+// used by tests to fake invoking the CLI
+type azureDeveloperCLITokenProvider func(ctx context.Context, scope []string, tenantID string) ([]byte, error)
+
+// AzureDeveloperCLICredentialOptions contains optional parameters for AzureDeveloperCLICredential.
+type AzureDeveloperCLICredentialOptions struct {
+	// AdditionallyAllowedTenants specifies tenants for which the credential may acquire tokens, in addition
+	// to TenantID. Add the wildcard value "*" to allow the credential to acquire tokens for any tenant the
+	// logged in account can access.
+	AdditionallyAllowedTenants []string
+	// TenantID identifies the tenant the credential should authenticate in.
+	// Defaults to the azd environment, which is the tenant where selected Azure subscription is.
+	TenantID string
+
+	tokenProvider azureDeveloperCLITokenProvider
+}
+
+// init returns an instance of AzureDeveloperCLICredentialOptions initialized with default values.
+func (o *AzureDeveloperCLICredentialOptions) init() {
+	if o.tokenProvider == nil {
+		o.tokenProvider = defaultAzdTokenProvider
+	}
+}
+
+// AzureDeveloperCLICredential authenticates as the identity logged in to the Azure Developer CLI.
```

Comment:
We should link to docs for people who don't know what azd is, or how to use it:
```suggestion
// AzureDeveloperCLICredential authenticates as the identity logged in to the [Azure Developer CLI].
//
// [Azure Developer CLI]: https://learn.microsoft.com/azure/developer/azure-developer-cli/overview
```